presto-main-base/src/test/java/com/facebook/presto/sql/planner/optimizations/TestAddExchangesPlansWithFunctions.java

```diff
@@ -0,0 +1,648 @@
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import org.testng.annotations.Test;
+
+import java.util.stream.Collectors;
+
+import static com.facebook.presto.common.type.TypeSignature.parseTypeSignature;
+import static com.facebook.presto.operator.scalar.annotations.ScalarFromAnnotationsParser.parseFunctionDefinitions;
+import static com.facebook.presto.spi.function.FunctionVersion.notVersioned;
+import static com.facebook.presto.spi.function.RoutineCharacteristics.Determinism.DETERMINISTIC;
+import static com.facebook.presto.spi.function.RoutineCharacteristics.Language.CPP;
+import static com.facebook.presto.spi.function.RoutineCharacteristics.Language.JAVA;
+import static com.facebook.presto.spi.function.RoutineCharacteristics.NullCallClause.RETURNS_NULL_ON_NULL_INPUT;
+import static com.facebook.presto.spi.plan.JoinType.INNER;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.anyTree;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.equiJoinClause;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.exchange;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.expression;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.filter;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.join;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.output;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.project;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.tableScan;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Scope.REMOTE_STREAMING;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Type.GATHER;
+import static com.facebook.presto.testing.TestingSession.testSessionBuilder;
+
+/**
+ * These are plan tests similar to what we have for other optimizers (e.g. {@link com.facebook.presto.sql.planner.TestPredicatePushdown})
+ * They test that the plan for a query after the optimizer runs is as expected.
+ * These are separate from {@link TestAddExchanges} because those are unit tests for
+ * how layouts get chosen.
+ * <p>
+ * Key behavior tested: When CPP functions are used with system tables, the filter containing
+ * the CPP function is preserved above the exchange (not pushed down) to ensure the filter
+ * executes in a different fragment from the system table scan. This validates the fragment
+ * boundary between CPP function evaluation and system table access.
+ */
+public class TestAddExchangesPlansWithFunctions
+        extends BasePlanTest
+{
+    public TestAddExchangesPlansWithFunctions()
+    {
+        super(TestAddExchangesPlansWithFunctions::createTestQueryRunner);
+    }
+
+    private static final SqlInvokedFunction CPP_FOO = new SqlInvokedFunction(
+            new QualifiedObjectName("dummy", "unittest", "foo"),
```

Comment:
nit: Renaming the function name from `foo` to `cpp_foo`, and similarly `bar` to `java_bar` will make gorking the tests a bit easier

Reply:
Great idea, done!